### DISCUSSION TITLE
Is the client supposed to be stripping undefined from return type?

### DISCUSSION THREAD

**Discussion by kswope** [trpc/trpc#5779] [Q&A]
I have this login proc here

```ts
 login: publicProc
    .input(z.object({ email: z.string(), password: z.string() }))
    .query(async (opts) => {
      const {
        input: { email, password },
      } = opts

      let user = await users.findByEmail(email)
      if (user) {
        let auth = await users.auth(user.id, password)
        if (auth) {
          let token = await logins.create(user.id)
          return { ok: true, token }
        }
      }

      return { ok: false, token: undefined }
    }),
```

The return type is indicated to be, which makes sense

```ts
output: {
        ok: boolean;
        token: string;
    } | {
        ok: boolean;
        token: undefined;
    };
```

If I use this in a test like it works correctly
```ts
// ...
let createCaller = createCallerFactory(router)
let client = createCaller({ authorization: undefined })
let { ok, token } = await client.public.login({ email, password })
...
```

However, when I use it in network scenario (express), I get a type error

```ts
let { ok, token } = await trpcClient.public.login.query({ email, password })
```

because the type has been changed to
```
        output: {
            ok: boolean;
            token: string;
        } | {
            ok: boolean;
        };
```

I'm aware undefined values get stripped out in JSON because they are not valid, but I'm not sure what is doing this.  Are the types being manipulated to account for this?.  If I change that procedure around I can get it to infer the return type of {ok: boolean, token: undefined|string}, but I shouldn't be expected to.

**Comment by KATT** [comment]
We have [a serializer](https://github.com/trpc/trpc/blob/next/packages/server/src/unstable-core-do-not-import/clientish/serialize.ts) that tries to match the typescript types to what JSON does

If you want to retain it, you can use [a transformer](https://trpc.io/docs/server/data-transformers)

**Reply by kswope** [reply]
Keeping

```
{ a: number | undefined }
```

but nuking

```
{ a: undefined }
```

is surprising behavior and doesn't seem like the right thing to do since it can be handled just fine without doing anything

I'm starting to question whether testing using server side calls is the right way to go since they aren't the same

**Reply by KATT** [reply]
```
{ a: number | undefined }
```

Shouldn't be nuked as it sometimes is defined -> ideal transform would be `{ a?: number }`, but ` | undefined` is the same for all intents and purposes

```
{ a: undefined }
```
This never shows up in a JSON payload and should be nuked.


Further, you likely want to make sure your `boolean` is a discriminated union to make it easier to work with - `as const` helps you with that


```ts
 login: publicProc
    .input(z.object({ email: z.string(), password: z.string() }))
    .query(async (opts) => {
      const {
        input: { email, password },
      } = opts

      let user = await users.findByEmail(email)
      if (user) {
        let auth = await users.auth(user.id, password)
        if (auth) {
          let token = await logins.create(user.id)
          return { ok: true as const, token }
        }
      }

      return { ok: false as const }
    }),
```

**Comment by kswope** [comment]
> This never shows up in a JSON payload and should be nuked.

disagree completely, since it does not matter, and you shouldn't be adding 'surprise' if it doesn't matter.

even the comments seem to agree with me in some way

```
// `undefined` is a weird one that's technically not valid JSON,
// but the return value of `JSON.parse` can be `undefined` so we
// support it as both a Primitive and a NonJsonPrimitive
```

appreciate this suggestion, but how many people even know this will result in a discriminated union?

```
      let user = await users.findByEmail(email)
      if (user) {
        let auth = await users.auth(user.id, password)
        if (auth) {
          let token = await logins.create(user.id)
          return { ok: true as const, token }
        }
      }

      return { ok: false as const }
```

That works, but in a couple of years, when I or somebody else looks at it, it will be not obvious what is going on.

But this is obvious and obvious to me is always better

```
      let user = await users.findByEmail(email)
      if (user) {
        let auth = await users.auth(user.id, password)
        if (auth) {
          let token = await logins.create(user.id)
          return { ok: true, token }
        }
      }

      return { ok: false, token: "" }

```
